src/hybridmap/HybridNeutralAtomMapper.cpp

```diff
@@ -0,0 +1,1293 @@
+  }
+
+  return distChange;
+}
+
+HwQubits NeutralAtomMapper::getBestMultiQubitPosition(const Operation* op) {
+  // try to find position around gate Qubits recursively
+  // if not, search through coupling graph until found according to a
+  // priority queue based on the distance to the other qubits
+
+  std::priority_queue<std::pair<fp, HwQubit>,
+                      std::vector<std::pair<fp, HwQubit>>, std::greater<>>
+      qubitQueue;
+  // add the gate qubits to the priority queue
+  auto gateQubits   = op->getUsedQubits();
+  auto gateHwQubits = this->mapping.getHwQubits(gateQubits);
+  // add the gate qubits to the priority queue
+  for (const auto& gateQubit : gateHwQubits) {
+    fp totalDist = 0;
+    for (const auto& otherGateQubit : gateHwQubits) {
+      if (gateQubit == otherGateQubit) {
+        continue;
+      }
+      totalDist +=
+          this->hardwareQubits.getSwapDistance(gateQubit, otherGateQubit, true);
+    }
+    qubitQueue.emplace(totalDist, gateQubit);
+  }
+
+  // run through the priority queue until a position is found
+  std::set<HwQubit> visitedQubits;
+  while (!qubitQueue.empty()) {
+    auto qubit = qubitQueue.top().second;
+    visitedQubits.insert(qubit);
+    qubitQueue.pop();
+
+    // remove selected qubit from the gate qubits
+    auto tempGateHwQubits = gateHwQubits;
+    if (tempGateHwQubits.find(qubit) != tempGateHwQubits.end()) {
+      tempGateHwQubits.erase(tempGateHwQubits.find(qubit));
+    }
+    auto bestPos = getBestMultiQubitPositionRec(
+        tempGateHwQubits, {qubit}, this->hardwareQubits.getNearbyQubits(qubit));
+    if (!bestPos.empty()) {
+      return bestPos;
+    }
+    // add nearby qubits to the priority queue
+    for (const auto& nearbyQubit :
+         this->hardwareQubits.getNearbyQubits(qubit)) {
+      if (visitedQubits.find(nearbyQubit) != visitedQubits.end()) {
+        continue;
+      }
+      // compute total distance to all other gate qubits
+      fp totalDist = 0;
+      for (const auto& otherGateQubit : gateHwQubits) {
+        if (nearbyQubit == otherGateQubit) {
+          continue;
+        }
+        totalDist +=
+            this->hardwareQubits.getSwapDistance(nearbyQubit, otherGateQubit);
+      }
+      qubitQueue.emplace(totalDist, nearbyQubit);
+    }
+  }
+  // find gate and move it to the shuttling layer
+  auto idxFrontGate =
+      std::find(this->frontLayerGate.begin(), this->frontLayerGate.end(), op);
+  if (idxFrontGate != this->frontLayerGate.end()) {
+    this->frontLayerGate.erase(idxFrontGate);
+    this->frontLayerShuttling.push_back(op);
+  }
+  // remove from lookahead layer if there
+  auto idxLookaheadGate = std::find(this->lookaheadLayerGate.begin(),
+                                    this->lookaheadLayerGate.end(), op);
+  if (idxLookaheadGate != this->lookaheadLayerGate.end()) {
+    this->lookaheadLayerGate.erase(idxLookaheadGate);
+    this->lookaheadLayerShuttling.push_back(op);
+  }
+  return {};
```

Comment:
All of this code is uncovered by tests. Is it actually needed? if yes, can it be tested?

Reply:
It actually might happen but these are edge cases which are really hard to construct by hand, so for the moment I would keep it as it is. Once I find a good test, I'll add it.

Reply:
I'll try to update the test such this edge case may occure